### ISSUE TITLE
error with PopupService

### ISSUE THREAD

**Issue by alyahmedaly** [ronzeidman/ng2-ui-auth#47]
https://github.com/ronzeidman/ng2-ui-auth/blob/master/src/oauth2.service.ts#L67 in this line you return oauthData returned from the server as it is in my case, it's throw exception because date type not observable i work around it by creating class that extend PopupService

```ts
import { PopupService } from 'ng2-ui-auth';
import { Observable } from 'rxjs/Observable';

/**
 * wrapper of Ng2UiAuthModule to handle our way in auth
 */
export class CustomPopupService extends PopupService {
    pollPopup() {
        return super.pollPopup().map(data => {
            if (data instanceof Observable) return data;
            return Observable.of(data);
        });
    };
}
```
here you can find the error 
```ts
TypeError: unknown type returned
    at Object.subscribeToResult (http://localhost:3000/vendor.bundle.js:10490:28)
    at MergeMapSubscriber._innerSub (http://localhost:3000/vendor.bundle.js:30879:39)
    at MergeMapSubscriber._tryNext (http://localhost:3000/vendor.bundle.js:30876:15)
    at MergeMapSubscriber._next (http://localhost:3000/vendor.bundle.js:30859:19)
    at MergeMapSubscriber.Subscriber.next (http://localhost:3000/vendor.bundle.js:10221:19)
    at DoSubscriber._next (http://localhost:3000/vendor.bundle.js:58225:31)
    at DoSubscriber.Subscriber.next (http://localhost:3000/vendor.bundle.js:10221:19)
    at TakeWhileSubscriber.nextOrComplete (http://localhost:3000/vendor.bundle.js:61259:26)
    at TakeWhileSubscriber._next (http://localhost:3000/vendor.bundle.js:61254:15)
    at TakeWhileSubscriber.Subscriber.next (http://localhost:3000/vendor.bundle.js:10221:19)
    at TakeSubscriber._next (http://localhost:3000/vendor.bundle.js:61049:31)
    at TakeSubscriber.Subscriber.next (http://localhost:3000/vendor.bundle.js:10221:19)
    at MergeMapSubscriber.notifyNext (http://localhost:3000/vendor.bundle.js:30892:31)
    at InnerSubscriber._next (http://localhost:3000/vendor.bundle.js:52466:22)
    at InnerSubscriber.Subscriber.next (http://localhost:3000/vendor.bundle.js:10221:19)
    at Object.subscribeToResult (http://localhost:3000/vendor.bundle.js:10447:26)
    at MergeMapSubscriber._innerSub (http://localhost:3000/vendor.bundle.js:30879:39)
    at MergeMapSubscriber._tryNext (http://localhost:3000/vendor.bundle.js:30876:15)
    at MergeMapSubscriber._next (http://localhost:3000/vendor.bundle.js:30859:19)
    at MergeMapSubscriber.Subscriber.next (http://localhost:3000/vendor.bundle.js:10221:19)
    at AsyncAction.IntervalObservable.dispatch (http://localhost:3000/vendor.bundle.js:54953:21)
    at AsyncAction._execute (http://localhost:3000/vendor.bundle.js:29070:19)
    at AsyncAction.execute (http://localhost:3000/vendor.bundle.js:29045:27)
    at AsyncScheduler.flush (http://localhost:3000/vendor.bundle.js:29142:33)
    at ZoneDelegate.invokeTask (http://localhost:3000/polyfills.bundle.js:7477:38)
    at Object.onInvokeTask (http://localhost:3000/vendor.bundle.js:6425:42)
    at ZoneDelegate.invokeTask (http://localhost:3000/polyfills.bundle.js:7476:43)
    at Zone.runTask (http://localhost:3000/polyfills.bundle.js:7377:48)
    at ZoneTask.invoke (http://localhost:3000/polyfills.bundle.js:7545:34)
```

**Comment by ronzeidman** [comment]
Hi, sorry for the late response, I've been pretty busy the last month.
Can you tell me what version of rxjs you have, I think beta version 9 and above automatically wrap the data and that's why I haven't had issues like this.

**Comment by ronzeidman** [comment]
I've tried to reproduce it, I cant see how. I've never expected the parameters returned from the popup to be an observable, they never are.

**Comment by alyahmedaly** [comment]
This my config and I will try to make plunder to simulate my use case 
auth config service
`@Injectable()
export class CustomAuthConfig extends CustomConfig {
    constructor({authservices, apiUrl}: ApiHelperService) {
        super();
        console.log(authservices);
        this.defaultHeaders = { 'Content-Type': 'application/json' };
        this.providers = {
            OWINGoogle: {
                url: null,
                authorizationEndpoint: `auth url`,
                requiredUrlParams: ['provider'],
                clientId: 'id',
                provider: 'Google',
                redirectUri: window.location.origin
            } as any
        };
    }
}`

popup service 
`import { PopupService } from 'ng2-ui-auth';
import { Observable } from 'rxjs/Observable';

/**
 * wrapper of Ng2UiAuthModule to handle our way in auth
 */
export class CustomPopupService extends PopupService {
    pollPopup() {
        return super.pollPopup().map(data => {
            if (data instanceof Observable) return data;
            return Observable.of(data);
        });
    };
}
`

call service with `this.uiAuth.authenticate('google')`
 sample of popup response 
`export interface OwinAuthResponse {
    external_access_token: string;
    provider: string;
    haslocalaccount: string | boolean;
    external_user_name: string;
}`

**Comment by alyahmedaly** [comment]
I don't need to return observable from popup service I return observable from it as hack,
As you see in this line [oauth2.service](https://github.com/ronzeidman/ng2-ui-auth/blob/master/src/oauth2.service.ts#L67) are returning the data that coming from popup service with concatmap I think concatmap is the one need data to be observable

**Comment by ronzeidman** [comment]
When you create a provider it's name is the provider key name and it's case-sensitive, so your provider should look like this:
```ts
{
            OWINGoogle: {
                /*not nessesary: url: null,*/
                authorizationEndpoint: `auth url`,
                requiredUrlParams: ['provider'],
                clientId: 'id',
                /*this does nothing: provider: 'Google',*/
                redirectUri: window.location.origin
            } as any
        }
```
and you call it like this:
`this.uiAuth.authenticate('OWINGoogle')`

I think I found the issue: https://github.com/ronzeidman/ng2-ui-auth/blob/b73dfc28a18b30cabb5d3399b24a932476d4ca7d/src/oauth2.service.ts#L67
this line should be `Observable.of(oauthData)`, can you try it and tell me if this fix works?

Thanks

**Comment by ronzeidman** [comment]
You can try install the master version: `npm i -S ronzeidman/ng2-ui-auth` and tell me if it works for you without workarounds.

**Comment by alyahmedaly** [comment]
yes it's now working as expected, thanks you 
but about this comment
`{
            OWINGoogle: {
                /*not nessesary: url: null,*/
                authorizationEndpoint: `auth url`,
                requiredUrlParams: ['provider'],
                clientId: 'id',
                /*this does nothing: provider: 'Google',*/
                redirectUri: window.location.origin
            } as any
        }`

 /*this does nothing: provider: 'Google'*/
my auth end point expecting provider in query string i work around it with 
                `requiredUrlParams: ['provider'],`
and yes i calling with `this.uiAuth.authenticate('OWINGoogle') `it was a typo

**Comment by alyahmedaly** [comment]
If we can provide better solution to append it's to url query string like key pair object
`requiredQueryStringParams:[{'provider':'Google'}] `

**Comment by alyahmedaly** [comment]
it will be great if I can pass url in config and inject [exchangeForToken](https://github.com/ronzeidman/ng2-ui-auth/blob/b73dfc28a18b30cabb5d3399b24a932476d4ca7d/src/oauth2.service.ts#L78)
function with my implementation if possible

**Comment by ronzeidman** [comment]
I thing this is an easy addition, I'll look into it next week.

**Comment by ronzeidman** [comment]
Took me a bit longer, try the master now (look at the https://github.com/ronzeidman/ng2-ui-auth/blob/master/CHANGELOG.md for more details)

**Comment by ronzeidman** [comment]
fixed in 4.1.0
